File: 4-median-of-two-sorted-arrays/4-median-of-two-sorted-arrays.cpp
class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        return solve(nums1, nums2);
    }
private:
    double solve(vector<int>& nums1, vector<int>& nums2) {
        if(nums2.size() < nums1.size())
            return solve(nums2, nums1);
        
        int n1 = nums1.size(), n2 = nums2.size();
        
        int low = 0, high = n1;
        int req = (n1 + n2 + 1) / 2;
        
        while(low <= high) {
            int mid = low + (high - low) / 2;
            int cut1 = mid;
            int cut2 = req - mid;
            
            int l1 = cut1 == 0 ? INT_MIN : nums1[cut1 - 1];
            int l2 = cut2 == 0 ? INT_MIN : nums2[cut2 - 1];
            int r1 = cut1 == n1 ? INT_MAX : nums1[cut1];
            int r2 = cut2 == n2 ? INT_MAX : nums2[cut2];
            
            if(l1 <= r2 && l2 <= r1) {
                if((n1 + n2) % 2 != 0)
                    return max(l1, l2);
                
                double res = max(l1, l2) + min(r1, r2);
                return res / 2;
            }
            if(l1 > r2) {
                high = mid - 1;
            }
            else
                low = mid + 1;
        }
        return 0.0;
    }
};

File: 4-median-of-two-sorted-arrays/4-median-of-two-sorted-arrays.py
class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        n1, n2 = len(nums1), len(nums2)
        if(n1 > n2):
            return self.findMedianSortedArrays(nums2, nums1)
        
        low, high = 0, n1
        req = (n1 + n2 + 1) // 2
        
        while low <= high:
            mid = low + (high - low) // 2
            cut1 = mid
            cut2 = req - cut1
            
            l1 = -float('inf') if cut1 == 0 else nums1[cut1 - 1]
            l2 = -float('inf') if cut2 == 0 else nums2[cut2 - 1]
            
            r1 = float('inf') if cut1 == n1 else nums1[cut1]
            r2 = float('inf') if cut2 == n2 else nums2[cut2]
            
            if l1 <= r2 and l2 <= r1:
                if (n1 + n2) % 2 == 0:
                    return (max(l1, l2) + min(r1, r2)) / 2
                return max(l1, l2)
            if l1 > l2:
                high = mid - 1
            else:
                low = mid + 1
        return 0.0